File: avocado/commands/sdk/dnf.py
"""SDK DNF command implementation."""

import os
import sys
import tomlkit


from avocado.commands.base import BaseCommand
from avocado.utils.container import SdkContainer
from avocado.utils.config import load_config
from avocado.utils.output import print_success, print_error
from avocado.utils.target import resolve_target, get_target_from_config


class SdkDnfCommand(BaseCommand):
    """Implementation of the 'sdk dnf' command."""

    @classmethod
    def register_subparser(cls, subparsers):
        """Register the sdk dnf command's subparser."""
        parser = subparsers.add_parser(
            "dnf", help="Run DNF commands in the SDK context"
        )

        # Add common arguments
        parser.add_argument(
            "-C",
            "--config",
            default="avocado.toml",
            help="Path to avocado.toml configuration file (default: avocado.toml)",
        )

        # Command is now required as a flag
        parser.add_argument(
            "-c",
            "--command",
            nargs="*",
            required=True,
            help="DNF command and arguments to execute",
        )

        return parser

    def execute(self, args, parser=None, unknown=None):
        """Execute the sdk dnf command."""
        # Get command args from -c/--command flag
        command_args = getattr(args, "command", [])
        if unknown:
            command_args.extend(unknown)

        if not command_args:
            print(
                "Error: No DNF command specified.",
                file=sys.stderr,
            )
            if parser:
                parser.print_help()
            return False

        config_path = args.config

        # Load configuration
        config, success = load_config(config_path)
        if not success:
            return False

        # Get the SDK image from configuration
        container_image = config.get("sdk", {}).get("image")
        if not container_image:
            print(
                "Error: No container image specified in config under 'sdk.image'",
                file=sys.stderr,
            )
            return False

        # Get repo_url from config, if it exists
        repo_url = config.get("sdk", {}).get("repo_url")
        repo_release = config.get("sdk", {}).get("repo_release")

        # Use resolved target (from CLI/env) if available, otherwise fall back to config
        config_target = get_target_from_config(config)
        target = resolve_target(
            cli_target=args.resolved_target, config_target=config_target
        )
        if not target:
            print(
                "Error: No target architecture specified. Use --target, AVOCADO_TARGET env var, or config under 'runtime.<name>.target'.",
                file=sys.stderr,
            )
            return False

        container_helper = SdkContainer()

        # Build dnf command
        command = f"RPM_CONFIGDIR=$AVOCADO_SDK_PREFIX/usr/lib/rpm $DNF_SDK_HOST $DNF_SDK_HOST_OPTS $DNF_SDK_REPO_CONF {
            ' '.join(command_args)}"

        # Run the DNF command using the container helper
        success = container_helper.run_in_container(
            container_image=container_image,
            target=target,
            command=[command],
            source_environment=False,
            use_entrypoint=True,
            repo_url=repo_url,
            repo_release=config.get("sdk", {}).get("repo_release"),
        )

        # Log the result
        if success:
            print_success(f"DNF command completed successfully.")
        else:
            print_error(f"DNF command failed.")

        return success

    def _print_help(self):
        """Print custom help message."""
        print("usage: avocado sdk dnf [-h] [-C CONFIG] -c COMMAND...")
        print()
        print("Run DNF commands in the SDK context")
        print()
        print("options:")
        print("  -h, --help            show this help message and exit")
        print("  -C CONFIG, --config CONFIG")
        print(
            "                        Path to avocado.toml configuration file (default: avocado.toml)"
        )
        print("  -c COMMAND, --command COMMAND")
        print("                        DNF command and arguments to execute (required)")
        print()
        print()
        print("Examples:")
        print("  avocado sdk dnf -c repolist --enabled")
        print("  avocado sdk dnf -c search python")
        print("  avocado sdk dnf -c install gcc make")
        print("  avocado sdk dnf -c --reinstall install cmake")
        print("  avocado sdk dnf -c list installed")

File: src/commands/sdk/dnf.rs
//! SDK DNF command implementation.

use anyhow::{Context, Result};

use crate::utils::{
    config::Config,
    container::{RunConfig, SdkContainer},
    output::{print_error, print_success, OutputLevel},
    target::resolve_target_required,
};

/// Implementation of the 'sdk dnf' command.
pub struct SdkDnfCommand {
    /// Path to configuration file
    pub config_path: String,
    /// Enable verbose output
    pub verbose: bool,
    /// DNF command and arguments to execute
    pub command: Vec<String>,
    /// Global target architecture
    pub target: Option<String>,
    /// Additional arguments to pass to the container runtime
    pub container_args: Option<Vec<String>>,
    /// Additional arguments to pass to DNF commands
    pub dnf_args: Option<Vec<String>>,
    /// SDK container architecture for cross-arch emulation
    pub sdk_arch: Option<String>,
}

impl SdkDnfCommand {
    /// Create a new SdkDnfCommand instance
    pub fn new(
        config_path: String,
        verbose: bool,
        command: Vec<String>,
        target: Option<String>,
        container_args: Option<Vec<String>>,
        dnf_args: Option<Vec<String>>,
    ) -> Self {
        Self {
            config_path,
            verbose,
            command,
            target,
            container_args,
            dnf_args,
            sdk_arch: None,
        }
    }

    /// Set SDK container architecture for cross-arch emulation
    pub fn with_sdk_arch(mut self, sdk_arch: Option<String>) -> Self {
        self.sdk_arch = sdk_arch;
        self
    }

    /// Execute the sdk dnf command
    pub async fn execute(&self) -> Result<()> {
        if self.command.is_empty() {
            return Err(anyhow::anyhow!(
                "No DNF command specified. Use --command (-c) to provide DNF arguments."
            ));
        }

        // Load the configuration
        let config = Config::load(&self.config_path)
            .with_context(|| format!("Failed to load config from {}", self.config_path))?;

        // Get the SDK image from configuration
        let container_image = config.get_sdk_image().ok_or_else(|| {
            anyhow::anyhow!("No container image specified in config under 'sdk.image'")
        })?;

        // Get repo_url and repo_release from config, if they exist
        let repo_url = config.get_sdk_repo_url();
        let repo_release = config.get_sdk_repo_release();

        // Merge container args from config with CLI args
        let merged_container_args = config.merge_sdk_container_args(self.container_args.as_ref());

        // Resolve target with proper precedence
        let target = resolve_target_required(self.target.as_deref(), &config)?;

        let container_helper = SdkContainer::new();

        // Build DNF command
        let dnf_args_str = if let Some(args) = &self.dnf_args {
            format!(" {} ", args.join(" "))
        } else {
            String::new()
        };
        let command = format!(
            "RPM_CONFIGDIR=$AVOCADO_SDK_PREFIX/usr/lib/rpm $DNF_SDK_HOST $DNF_SDK_HOST_OPTS $DNF_SDK_REPO_CONF --disablerepo=${{AVOCADO_TARGET}}-target-ext {} {}",
            dnf_args_str,
            self.command.join(" ")
        );

        // Run the DNF command using the container helper
        let success = self
            .run_dnf_command(
                &container_helper,
                container_image,
                &target,
                &command,
                repo_url.as_ref(),
                repo_release.as_ref(),
                merged_container_args.as_ref(),
            )
            .await?;

        // Log the result
        if success {
            print_success("DNF command completed successfully.", OutputLevel::Normal);
        } else {
            print_error("DNF command failed.", OutputLevel::Normal);
            return Err(anyhow::anyhow!("DNF command failed"));
        }

        Ok(())
    }

    /// Run DNF command using container with entrypoint
    #[allow(clippy::too_many_arguments)]
    async fn run_dnf_command(
        &self,
        container_helper: &SdkContainer,
        container_image: &str,
        target: &str,
        command: &str,
        repo_url: Option<&String>,
        repo_release: Option<&String>,
        container_args: Option<&Vec<String>>,
    ) -> Result<bool> {
        // Use the container helper's method with repo URL and release support
        let config = RunConfig {
            container_image: container_image.to_string(),
            target: target.to_string(),
            command: command.to_string(),
            verbose: self.verbose,
            source_environment: true, // need environment for DNF
            interactive: true,        // allow user input for DNF prompts
            repo_url: repo_url.cloned(),
            repo_release: repo_release.cloned(),
            container_args: container_args.cloned(),
            dnf_args: self.dnf_args.clone(),
            sdk_arch: self.sdk_arch.clone(),
            ..Default::default()
        };
        container_helper.run_in_container(config).await
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new() {
        let cmd = SdkDnfCommand::new(
            "config.toml".to_string(),
            false,
            vec!["install".to_string(), "gcc".to_string()],
            Some("test-target".to_string()),
            None,
            None,
        );

        assert_eq!(cmd.config_path, "config.toml");
        assert!(!cmd.verbose);
        assert_eq!(cmd.command, vec!["install", "gcc"]);
        assert_eq!(cmd.target, Some("test-target".to_string()));
    }

    #[tokio::test]
    async fn test_empty_command() {
        let cmd = SdkDnfCommand::new("config.toml".to_string(), false, vec![], None, None, None);

        let result = cmd.execute().await;
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("No DNF command specified"));
    }
}
